### PR TITLE
[ENH] `FourierFeatures`: more period options - `pandas` period alias and from offset column

### LINKED ISSUE #5512: [ENH] FourierFeatures more periodicity options
## Is your feature request related to a problem? Please describe.

The current implementation of Sktime's FourierFeatures states the following:

> 
> Fourier Series terms can be used as explanatory variables for the cases of multiple
> seasonal periods and or complex / long seasonal periods [1]_, [2]_. For every
> seasonal period, $sp$ and fourier term $k$ pair there are 2 fourier
> terms sin_sp_k and cos_sp_k:
> - sin_sp_k = $sin(\frac{2 \pi k t}{sp})$
> - cos_sp_k = $cos(\frac{2 \pi k t}{sp})$
> 
> Where $t$ is the number of time steps elapsed from the beginning of the time
> series.

> Parameters
> 
> sp_list : List[float]
> - list of seasonal periods

This implementation however poses a serious limitation when the user is interested in periodicity that occurs with different interval lengths. Some examples:
- monthly
    - can differ in length due to different number of days per month
- yearly
    - can differ in length due to leap year

## Describe the solution you'd like
In addition to expressing periodicity as the number of timesteps, we give the user the option to define periodicity as a pandas period alias or as a column name in the input data. The enhanced docstring is as follows:

> 
> Fourier Series terms can be used as explanatory variables for the cases of multiple
> seasonal periods and or complex / long seasonal periods [1]_, [2]_. For every
> seasonal period, $sp$ and fourier term $k$ pair there are 2 fourier
> terms sin_sp_k and cos_sp_k:
> - sin_sp_k = $sin(\frac{2 \pi k t}{sp})$
> - cos_sp_k = $cos(\frac{2 \pi k t}{sp})$
> 
> Where $t$ is the elapsed time since the beginning of the seasonal period and
> $sp$ the total time of the seasonal period.

> Parameters
> 
> sp_list : List[float or str]
> - List of seasonal periods. A float defines the length of the seasonality starting at the beginning of the time series in number of timesteps. A string should match the column name in X that contains the $t/sp$ values or should match a pandas period alias: https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#period-aliases

See the #5513 for the implementation of the proposed solution.

## Describe alternatives you've considered
None, as the proposed solution:
- is easy to understand for the user
- offers great flexibility for the user
- does not change the current functionality in any way

### PR DESCRIPTION
**Pull request by Ram0nB**
#### Reference Issues/PRs
Fixes #5512

#### What does this implement/fix? Explain your changes.

See Issue #5512 for a description of the problem and the solution implemented in this PR.

For periodicity defined as a pandas period alias, a private method is added that takes the period alias and a datetime index, and returns per datetime the fraction of the period that has passed in the interval [0, 1).

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
All changed files :)

#### Did you add any tests for the change?
No

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

File: sktime/transformations/series/fourier.py [hunk 1]
```diff
@@ -23,23 +23,27 @@ class FourierFeatures(BaseTransformer):
...
+    :math:`sp` the total time of the seasonal period.
 
-    The output of the transform is a pandas DataFrame that includes the fourier terms as
+    The transformed output is a pandas DataFrame that includes the fourier terms as
     additional columns with the naming convention stated above (sin_sp_k and cos_sp_k).
-    For instance for sp_list = [12, 3] and fourier_terms_list = [2, 1] the transformed
+    For instance for sp_list = [12, "Y"] and fourier_terms_list = [2, 1] the transformed
     series will have the additional columns:
-    "cos_12_1", "sin_12_1", "cos_12_2", "sin_12_2", "cos_3_1", "sin_3_1"
+    "cos_12_1", "sin_12_1", "cos_12_2", "sin_12_2", "cos_Y_1", "sin_Y_1"
```
Comment (fkiraly) [inline sktime/transformations/series/fourier.py:33]: something isn't quite right here. Is this the paired product or the cross-product?

If we pair 12, Y and 2,1, then we get (12, 2) and (Y, 1), so should we not get `"cos_12_2", "sin_12_2", "cos_Y_1", "sin_Y_1"` only?

Whereas if we cross-product them, we get (12, 2), (12, 1), (Y, 2), (Y, 1), so should we not get `"cos_12_2", "sin_12_2", "cos_12_1", "sin_12_1", "cos_Y_2", "sin_Y_2", "cos_Y_1", "sin_Y_1"`, in this order?

Neither plausible options seems to give the claimed columns.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice!

Blocking, perhaps just an explanation: which columns exactly does this produce? The issue already seems to have been with the estimator as is, but we still should fix the documentation to agree with the actual.

Non-blocking: there are now multiple options for `sp_list`, I would translate the dense paragraph to a bullet point list of options.

**Comment by Ram0nB** [comment]
> Nice!
> 
> Blocking, perhaps just an explanation: which columns exactly does this produce? The issue already seems to have been with the estimator as is, but we still should fix the documentation to agree with the actual.
> 
> Non-blocking: there are now multiple options for `sp_list`, I would translate the dense paragraph to a bullet point list of options.

I agree with both points: I just updated the docstring. Let me know if you think this is clear enough for the user :smiley:

**Review by fkiraly — CHANGES_REQUESTED** [review]
Ah, it's paired and "up to the number".

Although the math is a bit discrepant, the math claims that the indexing is 0-based, but the naming is 1-based.

Can you kindly check?

File: sktime/transformations/series/fourier.py (hunk 1 above)
Comment (fkiraly) [reply sktime/transformations/series/fourier.py:40]: thanks for the explainer, clear now.

**Review by fkiraly — APPROVED** [review]
Updated the docstring, so it's ready for the release.

**Pull request merged**